### DISCUSSION TITLE
The correct way to test custom resolver

### DISCUSSION THREAD

**Discussion by evgenoir** [mswjs/msw#1253] [Q&A] [answered]
I've created some custom response resolver

```js
export const withBearerAuthorization = (resolver) => {
    return (req, res, ctx) => {
        const authorization = req.headers.get("authorization")
        if (!authorization || !authorization.startsWith("Bearer "))
            return res(
                ctx.status(200),
                ctx.json({
                    msg: "Missing Authorization Header"
                })
            )

        return resolver(req, res, ctx)
    }
}
```

and test it like this

```js
const stubResolver = (req, res, ctx) => res(
    ctx.status(200)
)

describe("withBearerAuthorization tests", () => {
    test("Request without authorization header", async () => {
        const requestWithoutAuthHeader = new Request('/some_url')

        const result = await withBearerAuthorization(stubResolver)(requestWithoutAuthHeader, response, context)

        // Result has the following format:
        // {
        //     status: 401,                                    // Number
        //     body: "{"msg":"Missing Authorization Header"}"  // String
        //     ...                                             // other properties
        // }
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('body')
        expect(result.body).toStrictEqual(JSON.stringify({
            msg: "Missing Authorization Header"
        }))
    })
})
```

My question may seem a little strange, but I'm looking for **a better approach**: Is this the correct way to test custom resolver, or is there a better approach for testing it?
Any improvements are welcome :)

**Comment by kettanaito** [comment, marked as answer]
Hi, @evgenoir. That's a perfect question, thanks for raising this. 

Overall, your testing direction is solid. I will try adding a few points below.

## `Request` vs `MockedRequest`

I like that you construct a plain `Request` instance to pass as the `req` argument. But, strictly speaking, the `req` we construct in MSW is not a 1-1 equivalent of the fetch's `Request`, as the library's one has extra properties (like `passthrough()` or `params`). You may run into some issues as you're passing in not what MSW would produce on its own. 

We currently don't expose the means to construct a `MockedRequest` instance publicly as we treat that instance as an implementation detail of the library. For testing purposes internally we have a `createMockedResponse` utility:

https://github.com/mswjs/msw/blob/6cfe67ffd846cee2c311bc7239478230afe99685/test/support/utils.ts#L19-L44

You can get inspired by this and introduce something similar to your codebase. This would work best with TypeScript so you can catch changes to `MockedRequest` between versions. 

## Abstracting the resolver

The `x()()` call signature may get confusing for some people so I can recommend abstracting the actual resolver to a separate variable.

```js
const resolver = withBearerAuthorization(stubResolver)

const response = await resolver(requestWithoutAuthorization, response, context)
```

Even more: once you abstract the `resolver`, you notice that it'd stay the same for all your test because you're not testing `stubResolver` in any way, but instead assert implementation of the `withBearerAuthorization` given a certain input. 

## Context

I think you're doing this right already but wanted to mention that you need to use `restContext` as the value for the `ctx` argument:

```js
import { restContext } from 'msw'

resolver(requestWithoutAuthorization, response, restContext)
```

This will ensure that your higher-order resolver indeed has `ctx.json` and friends to execute. 

## Other

You may rename `result` to `response` because that's what a resolver returns: a `MockedResponse` representation.

**Reply by evgenoir** [reply]
Great. This is exactly what I was looking for. Thanks!

**Reply by kettanaito** [reply]
Happy to hear that helped!
I like helping people wherever I can. If you'd like to support me, and the effort behind MSW, please consider [sponsoring the project](https://github.com/sponsors/mswjs). Your contribution would mean a lot to me. Thank you!
